library/src/main/java/com/novoda/downloadmanager/RoomAppDatabase.java

```diff
@@ -8,24 +8,33 @@
 @Database(entities = {RoomBatch.class, RoomFile.class}, version = 1)
 abstract class RoomAppDatabase extends RoomDatabase {
 
-    private static volatile RoomAppDatabase SINGLE_INSTANCE;
+    private static volatile RoomAppDatabase singleInstance;
+    private static final Object LOCK = new Object();
 
     abstract RoomBatchDao roomBatchDao();
 
     abstract RoomFileDao roomFileDao();
 
     static RoomAppDatabase obtainInstance(Context context) {
-        if (SINGLE_INSTANCE == null) {
-            SINGLE_INSTANCE = newInstance(context);
+        RoomAppDatabase database = singleInstance;
+        if (database == null) {
+            synchronized (LOCK) {
+                database = singleInstance;
+                if (database == null) {
+                    singleInstance = newInstance(context);
+                    database = singleInstance;
```

Comment:
Lines 19, 22 and 25 are the same. The previous solution was functionally the same as this but clearer to read and didn't need a lock.

Reply:
@joetimmins I linked this https://www.journaldev.com/1377/java-singleton-design-pattern-best-practices-examples#thread-safe-singleton which suggests that both solutions are very different. I'm happy to change it back if you can explain why this is the same?

Reply:
Static analysis complains about `synchronized` methods as opposed to static blocks because they are a performance hit.

Reply:
Saying that I can find a number of other articles that show the `double checked locking` mechanism is stupid too 😂  just can't win, whatever we do 😄

Reply:
was it functionally the same? with the new code only 1 instance of the database will be created but with the old one it'd be possible that 2 threads could create 2 instances because the synchronisation was only on the `newInstance` method, so the 2 threads could execute the null check `if (SINGLE_INSTANCE == null)` and then 1 of them runs `newInstance` while the other waits, once that completes the other thread executes `newInstance` again - correct me if I'm wrong but I think the new solution while less readable it is more correct (in terms of muti threading safety)

Reply:
@xrigau - doesn't the `volatile` keyword prevent this?

Reply:
This is not a hyperthreading program and there is not a major issue if the migration and the downloader at some point have two different instances of the database, specially when the database has transactions.

Reply:
Yea that was my understanding too @xrigau 

I think though that @joetimmins is right about the readability and we should be able to achieve the same with:

```
     if (singleInstance == null) {
            synchronized (RoomAppDatabase.class) {
                if (singleInstance == null) {
                    singleInstance = newInstance(context);
                }
            }
        }
        return singleInstance;
```

Reply:
@joetimmins it solves in a different thread execution but not when two threads enters the `newInstance` in parallel. One thread will be blocked and the other will start, when it finishes, the blocked thread will enter and create the instance again; the double null check prevents that scenario and also only that method is blocked and not the entire class due to the synchronised method.

Reply:
@Mecharyry the blog post you linked says that having the method containing the null check marked as `synchronized` is a performance hit, because it is called every time we obtain an instance. In the previous solution here, only the object instantiation was `synchronized`, the null check was not. This was because the single instance was marked as `volatile`, meaning it has to be retrieved from main memory every time it is read, and written to main memory whenever it is written, preventing other threads from having out of date copies of it. Which I think mitigates @xrigau 's concern too.

If this is wrong, please tell me what is actually happening here instead 😁

Reply:
this wikipedia article explains the differences between the solution in this PR and the code in your comment https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java but they're actually slightly different, regarding the `volatile` keyword I'm actually not sure I thought it only affected the order of reads and writes but not the synchronisation between threads and shared memory

Reply:
The updated code seems to match the solution you linked now @xrigau 😄  I think this is the neatest solution so far but I don't know if I have missed anything 😄